src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs

```diff
@@ -104,6 +104,29 @@ protected override async Task OnMessageReceivedAsync(INetworkPeer peer, Incoming
             }
         }
 
+        /// <inheritdoc />
+        protected override bool CanConsumeCache()
+        {
+            int height = this.consensusManager.Tip.Height;
+
+            // Try cashe consumption every N blocks advanced by consensus.
+            // N should be between 0 and max reorg. 20% of max reorg is a good random value.
+            // Higher the N the better performance boost we can get.
+            uint tryCacheEveryBlocksCount = this.network.Consensus.MaxReorgLength / 5;
+
+            if (height == this.lastCheckpointHeight)
+                return true;
+
+            if (height > this.lastCheckpointHeight)
+            {
+                // After last checkpoint.
+                if ((height % tryCacheEveryBlocksCount == 0) || (height >= this.BestReceivedTip.Height))
```

Comment:
Is there a possibility BestReceivedTip can be null? (probably not as first run headers will always be found)

Reply:
How can this be possible?
`height >= this.BestReceivedTip.Height` perhaps equal to is possible  (so I guess for sanity also check greater then)

Reply:
>Is there a possibility BestReceivedTip can be null? (probably not as first run headers will always be found)

Good spot. Yeah, I think it's possible. 
Fixed

Reply:
>How can this be possible?
height >= this.BestReceivedTip.Height perhaps equal to is possible (so I guess for sanity also check greater then)

It's possible when peer is behind our consensus.

Reply:
ah yeah sure in case the peer is in a fork (within the max reorg) and trying to connect a longer chain it will be under tip indeed very well.